File: homework4/design_pattern/csharp/adapter.cs
using System;

// Internal class with a Fetch method
public class InternalClass
{
    public string Fetch()
    {
        return "get user info";
    }
}

// External class with a Search method
public class ExternalClass
{
    public string Search()
    {
        return "get user info";
    }
}

// Adapter class to adapt ExternalClass to match InternalClass's interface
public class Adapter
{
    private readonly ExternalClass _external;

    public Adapter(ExternalClass external)
    {
        _external = external;
    }

    public string Fetch()
    {
        return _external.Search();
    }
}

// Main program
class Program
{
    static void Main(string[] args)
    {
        // Create an instance of ExternalClass
        ExternalClass external = new ExternalClass();

        // Use the Adapter to adapt ExternalClass to match InternalClass's interface
        Adapter adapter = new Adapter(external);

        // Call the Fetch method through the adapter
        Console.WriteLine(adapter.Fetch()); // Output: get user info
    }
}

File: homework4/design_pattern/cpp/adapter.cpp
#include <iostream>
#include <string>

// InternalClass
class InternalClass {
public:
    std::string fetch() {
        return "get user info";
    }
};

// ExternalClass
class ExternalClass {
public:
    std::string search() {
        return "get user info";
    }
};

// Adapter
class Adapter {
private:
    ExternalClass* external;
public:
    Adapter(ExternalClass* external) : external(external) {}

    std::string fetch() {
        return external->search();
    }
};

int main() {
    ExternalClass external;
    Adapter adapter(&external);

    std::cout << adapter.fetch() << std::endl;

    return 0;
}